### PR TITLE
Support specifying which IP addresses to listen on

### PR DESCRIPTION
**Pull request by agners**
Allow to define which IP addresses to bind the WebSocket server on. This is especially useful to limit which addresses the add-on will listen on since it runs in host network context.

### CODE REVIEW

File: matter_server/server/server.py [hunk 1]
```diff
@@ -62,13 +62,15 @@ def __init__(
         vendor_id: int,
         fabric_id: int,
         port: int,
+        listen_addresses: list[str],
```
Comment (MartinHjelmare) [inline matter_server/server/server.py:65]:
```suggestion
        listen_addresses: list[str] | None = None,
```

File: matter_server/server/server.py [hunk 2]
```diff
@@ -62,13 +62,15 @@ def __init__(
...
         primary_interface: str | None,
```
Comment (MartinHjelmare) [inline matter_server/server/server.py:66]:
```suggestion
        primary_interface: str | None = None,
```

File: matter_server/server/server.py [hunk 3]
```diff
@@ -62,13 +62,15 @@ def __init__(
...
-        primary_interface: str | None,
+        listen_addresses: list[str] | None = None,
+        primary_interface: str | None = None,
     ) -> None:
         """Initialize the Matter Server."""
         self.storage_path = storage_path
         self.vendor_id = vendor_id
         self.fabric_id = fabric_id
-        self.port = port
+        self._port = port
```
Comment (MartinHjelmare) [inline matter_server/server/server.py:72]: Was this change intended? Technically a breaking change but the attribute is probably not needed for outside use, besides for debugging.

Comment (agners) [reply matter_server/server/server.py:72]: My thought was to align it with `listen_address`, which I preferred to be private. I could a read only property instead? :thinking:

Comment (marcelveldt) [reply matter_server/server/server.py:72]: I'd leave it public

File: matter_server/server/server.py [hunk 4]
```diff
@@ -62,13 +62,15 @@ def __init__(
         vendor_id: int,
         fabric_id: int,
         port: int,
-        primary_interface: str | None,
+        listen_addresses: list[str] | None = None,
```
Comment (marcelveldt) [inline matter_server/server/server.py:65]: is it indended to have a list here ?

File: matter_server/server/__main__.py [hunk 5]
```diff
@@ -95,6 +104,7 @@ def main() -> None:
         int(args.vendorid),
         int(args.fabricid),
         int(args.port),
+        args.listen_address,
```
Comment (marcelveldt) [inline matter_server/server/__main__.py:107]: you are sending a None or single string value here while the typing of "listen_addresses" is set to a list of strings or None.

File: matter_server/server/server.py (hunk 4 above)
Comment (agners) [reply matter_server/server/server.py:65]: Yes, you may listen on multiple addresses or address families. E.g. to explicitly listen to what is the default today you'd pass `--listen-address 0.0.0.0 --listen-address ::`.

File: matter_server/server/__main__.py (hunk 5 above)
Comment (marcelveldt) [reply matter_server/server/__main__.py:107]: nvm, I overlooked the append.

File: matter_server/server/server.py (hunk 4 above)
Comment (marcelveldt) [reply matter_server/server/server.py:65]: Yeah, see above, I overlooked the append. All good.

**Review by marcelveldt — APPROVED** [review]

**Comment by marcelveldt** [comment]
see the linter error from mypy

**Comment by agners** [comment]
> see the linter error from mypy

Ugh, yeah I see `TCPSite` supports only a single string as host, whereas the underlying `loop.create_server` supports a list. We've run into this limitation in the Core already (see https://github.com/home-assistant/core/blob/2024.2.0b8/homeassistant/components/http/web_runner.py#L14).

I've now added a similar implementation here too.

@bdraco do you think we can add something like that to aiohttp? :pleading_face:

**Review by marcelveldt — APPROVED** [review]
